expfmt/expfmt.go

```diff
@@ -14,30 +14,121 @@
+	OpenMetricsVersion_2_0_0 = "2.0.0"
 	OpenMetricsVersion_1_0_0 = "1.0.0"
+	OpenMetricsVersion_0_0_1 = "0.0.1"
 
-	// The Content-Type values for the different wire protocols.
+	// The Content-Type values for the different wire protocols. Do not do direct
+	// to comparisons to these constants, instead use the comparison functions.
 	FmtUnknown           Format = `<unknown>`
-	FmtText              Format = `text/plain; version=` + TextVersion + `; charset=utf-8`
+	FmtText_0_0_4        Format = `text/plain; version=` + TextVersion_0_0_4 + `; charset=utf-8`
+	FmtText_1_0_0        Format = `text/plain; version=` + TextVersion_1_0_0 + `; charset=utf-8`
 	FmtProtoDelim        Format = ProtoFmt + ` encoding=delimited`
 	FmtProtoText         Format = ProtoFmt + ` encoding=text`
 	FmtProtoCompact      Format = ProtoFmt + ` encoding=compact-text`
-	FmtOpenMetrics_1_0_0 Format = OpenMetricsType + `; version=` + OpenMetricsVersion_1_0_0 + `; charset=utf-8`
 	FmtOpenMetrics_0_0_1 Format = OpenMetricsType + `; version=` + OpenMetricsVersion_0_0_1 + `; charset=utf-8`
+	FmtOpenMetrics_1_0_0 Format = OpenMetricsType + `; version=` + OpenMetricsVersion_1_0_0 + `; charset=utf-8`
+	FmtOpenMetrics_2_0_0 Format = OpenMetricsType + `; version=` + OpenMetricsVersion_2_0_0 + `; charset=utf-8`
+
+	// UTF8 and Escaping Formats
+	FmtUTF8Param         Format = `; validchars=utf8`
 )
 
 const (
 	hdrContentType = "Content-Type"
 	hdrAccept      = "Accept"
 )
+
+type FormatType int
+
+const (
+	TypeUnknown = iota
+	TypeProtoCompact
+	TypeProtoDelim
+	TypeProtoText
+	TypeTextPlain
+	TypeOpenMetrics
+)
+
+func (f Format) ContentType() FormatType {
+	toks := strings.Split(string(f), ";")
+	if len(toks) < 2 {
+		return TypeUnknown
+	}
+
+	params := make(map[string]string)
+	for i, t := range toks {
+		if i == 0 {
+			continue
+		}
+		args := strings.Split(t, "=")
+		if len(args) != 2 {
+			continue
+		}
+		params[strings.TrimSpace(args[0])] = strings.TrimSpace(args[1])
+	}
+
+	switch strings.TrimSpace(toks[0]) {
+	case ProtoType:
+		if params["proto"] != ProtoProtocol {
+			return TypeUnknown
+		}
+		switch params["encoding"] {
+		case "delimited":
+			return TypeProtoDelim
+		case "text":
+			return TypeProtoText
+		case "compact-text":
+			return TypeProtoCompact
+		default:
+			return TypeUnknown
+		}
+	case OpenMetricsType:
+		if params["charset"] != "utf-8" {
+			return TypeUnknown
+		}
+		return TypeOpenMetrics
+	case "text/plain":
+		v, ok := params["version"]
+		if !ok {
+			return TypeTextPlain
+		}
+		if v == TextVersion_0_0_4 || v == TextVersion_1_0_0 {
+			return TypeTextPlain
+		}
+		return TypeUnknown
+	default:
+		return TypeUnknown
+	}
+}
+
+func (format Format) SupportsUTF8() bool {
+	for _, p := range strings.Split(string(format), ";") {
+		toks := strings.Split(p, "=")
+		if len(toks) != 2 {
+			continue
+		}
+		key, value := strings.TrimSpace(toks[0]), strings.TrimSpace(toks[1])
+		// By definition, if utf8 is allowed then names are not escaped.
```

Comment:
this comment is leftover cruft from a previous version of this function and can be removed